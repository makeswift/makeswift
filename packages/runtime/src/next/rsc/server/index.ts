import 'server-only'

export { NextRSCMakeswiftPage as ExperimentalMakeswiftPage } from './makeswift-page'
export { NextRSCServerProvider as ExperimentalServerProvider } from './server-provider'
