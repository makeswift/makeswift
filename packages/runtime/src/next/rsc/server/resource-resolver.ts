import { FetchableValue, ResourceResolver } from '@makeswift/controls'
import { type SiteVersion } from '../../../api/site-version'
import { ReactRuntimeCore } from '../../../runtimes/react/react-runtime-core'
import { Document, getElementId } from '../../../state/react-page'
import { MakeswiftClient } from '../../../client'

function createFetchable<T>(
  name: string,
  fetcher: () => Promise<T | null>,
): FetchableValue<T | null> {
  let lastValue: T | null = null

  return {
    name,
    subscribe: () => () => {},
    readStable: () => lastValue,
    fetch: async () => {
      const res = await fetcher()
      lastValue = res
      return res
    },
  }
}

export function serverResourceResolver(
  runtime: ReactRuntimeCore,
  client: MakeswiftClient,
  siteVersion: SiteVersion | null,
  { key: documentKey, locale }: Document,
): ResourceResolver {
  const serverResourceResolver: ResourceResolver = {
    resolveSwatch(swatchId: string | undefined) {
      return createFetchable(`swatch:${swatchId}`, async () => {
        if (!swatchId) return null
        return client.getSwatch(swatchId, siteVersion)
      })
    },

    resolveFile(fileId: string | undefined) {
      return createFetchable(`file:${fileId}`, async () => {
        if (!fileId) return null
        return client.getFile(fileId)
      })
    },

    resolveTypography(typographyId: string | undefined) {
      return createFetchable(`typography:${typographyId}`, async () => {
        if (!typographyId) return null
        return client.getTypography(typographyId, siteVersion)
      })
    },

    resolvePagePathnameSlice(pageId: string | undefined) {
      return createFetchable(`pagePathnameSlice:${pageId}`, async () => {
        if (!pageId) return null
        return client.getPagePathnameSlice(pageId, siteVersion, {
          locale: locale ?? undefined,
        })
      })
    },

    resolveElementId(elementKey: string) {
      return {
        name: `elementId:${documentKey}:${elementKey}`,
        subscribe: () => () => {},
        readStable: () => getElementId(runtime.store.getState(), documentKey, elementKey),
        fetch: async () => null,
      }
    },
  }

  return serverResourceResolver
}
