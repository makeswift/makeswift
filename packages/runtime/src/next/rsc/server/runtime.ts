import 'server-only'

import { cache } from 'react'
import { ReactRuntime } from '../../../runtimes/react/react-runtime'
import { SiteVersion } from '../../../unstable-framework-support'

const getRuntimeCache = cache((): { runtime: ReactRuntime | null } => ({ runtime: null }))
const getSiteVersionCache = cache((): { siteVersion: SiteVersion | null } => ({
  siteVersion: null,
}))

export function setRuntime(runtime: ReactRuntime): void {
  const cache = getRuntimeCache()
  cache.runtime = runtime
}

export const getRuntime = (): ReactRuntime => {
  const cache = getRuntimeCache()

  if (!cache.runtime) {
    throw new Error('Makeswift context not found. Ensure your page uses `runWithMakeswiftContext`.')
  }
  return cache.runtime
}

export function setSiteVersion(siteVersion: SiteVersion | null): void {
  const cache = getSiteVersionCache()
  cache.siteVersion = siteVersion
}

export function getSiteVersionFromCache(): SiteVersion | null {
  const cache = getSiteVersionCache()
  return cache.siteVersion
}
