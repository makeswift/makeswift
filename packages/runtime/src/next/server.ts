export { MakeswiftApiHandler } from './api-handler'
export type { Manifest, Font } from '../api-handler'
export type { MakeswiftApiHandlerResponse } from './api-handler'
export { getSiteVersion } from './draft-mode'
