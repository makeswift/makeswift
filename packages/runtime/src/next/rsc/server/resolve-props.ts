import { ElementData, StyleDefinition } from '@makeswift/controls'
import {
  getDocumentFromCache,
  getMakeswiftClient,
  getRuntime,
  getSiteVersionFromCache,
} from './runtime'
import { getBreakpoints } from '../../../state/react-page'
import { createCollectingServerStylesheet } from '../css/server-css'
import { isLegacyDescriptor } from '../../../prop-controllers/descriptors'
import { DescriptorsByProp } from '../../../state/modules/prop-controllers'
import { serverResourceResolver } from './resource-resolver'

export async function resolveProps(
  element: ElementData,
  propDescriptors: DescriptorsByProp,
): Promise<Record<string, unknown>> {
  const runtime = getRuntime()
  const siteVersion = getSiteVersionFromCache()
  const document = getDocumentFromCache()
  const client = getMakeswiftClient()

  const state = runtime.store.getState()
  const breakpoints = getBreakpoints(state)
  const stylesheet = createCollectingServerStylesheet(breakpoints, element.key)

  const serverResolver = serverResourceResolver(runtime, client, siteVersion, document)

  const resolvedProps = await Promise.all(
    Array.from(Object.entries(propDescriptors)).map(async ([propName, descriptor]) => {
      if (isLegacyDescriptor(descriptor)) {
        console.warn(`[resolveProps] Cannot use legacy descriptor in RSC. Prop: ${propName}`)
        return [propName, undefined] as const
      }

      const propData = element.props[propName]
      const isStyleControl = descriptor.controlType === StyleDefinition.type

      // Always process style controls, even when they have no data
      if (propData !== undefined || isStyleControl) {
        const resolvedValue = descriptor.resolveValue(
          propData,
          serverResolver,
          stylesheet.child(propName),
        )
        // TODO: This is a hack for style controls to have their styles injected
        // on the server. CSS injection on the server appears to have broken
        // when introducing the await resolvedValue.triggerResolve() ahead of
        // the `readStable` call â€” for proof, comment out the triggerResolve and
        // observe that styles are correct on load.
        resolvedValue.readStable()
        await resolvedValue.triggerResolve()
        const result = resolvedValue.readStable()
        return [propName, result] as const
      }

      return [propName, undefined] as const
    }),
  )

  const resolvedPropsRecord: Record<string, unknown> = Object.fromEntries(resolvedProps)
  return resolvedPropsRecord
}
