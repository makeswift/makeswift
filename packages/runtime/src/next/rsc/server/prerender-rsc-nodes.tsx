import { getRuntime } from './runtime'
import { getComponentsMeta } from '../../../state/modules/components-meta'
import { Element, getPropControllerDescriptors } from '../../../state/react-page'
import { ServerElement } from './server-element'
import { traverseElementTree } from '../../../state/modules/element-trees'
import { RSCNodes } from '../client/rsc-nodes-provider'

export function prerenderRSCNodes(elementTree: Element): RSCNodes {
  const runtime = getRuntime()
  const state = runtime.store.getState()
  const descriptors = getPropControllerDescriptors(state)
  const rscNodes: RSCNodes = new Map()

  for (const element of traverseElementTree(elementTree, descriptors)) {
    const meta = getComponentsMeta(state.componentsMeta).get(element.type)

    if (meta == null) {
      console.warn(`[prerenderRSCNodes] Component meta not found for ${element.type}`)
      continue
    }

    if (rscNodes.has(element.key)) {
      console.warn(`[prerenderRSCNodes] RSC node already exists for ${element.key}`)
      continue
    }

    if (meta.server) {
      rscNodes.set(element.key, <ServerElement key={element.key} element={element} />)
    }
  }

  return rscNodes
}
