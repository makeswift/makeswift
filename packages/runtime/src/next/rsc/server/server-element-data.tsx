import {
  getReactComponent,
  ElementData,
  getComponentPropControllerDescriptors,
} from '../../../state/react-page'
import { FallbackComponent } from '../../../components/shared/FallbackComponent'
import { getRuntime } from './runtime'
import { resolveProps } from './resolve-props'

type Props = {
  elementData: ElementData
}

export async function ServerElementData({ elementData }: Props): Promise<JSX.Element> {
  const state = getRuntime().store.getState()
  const Component = getReactComponent(state, elementData.type)

  let descriptors = getComponentPropControllerDescriptors(state, elementData.type)

  if (descriptors == null) {
    return <FallbackComponent text={`Descriptors not found for ${elementData.type}`} />
  }

  const props = await resolveProps(elementData, descriptors)

  if (Component == null) {
    return <FallbackComponent text="Component not found" />
  }

  return <Component {...props} key={elementData.key} />
}
