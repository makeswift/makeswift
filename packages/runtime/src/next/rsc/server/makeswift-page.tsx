import { ComponentPropsWithoutRef } from 'react'
import { prerenderRSCNodes } from './prerender-rsc-nodes'
import { RSCNodesProvider } from '../client/rsc-nodes-provider'
import { CSSInjector } from '../css/server-css'
import { ClientCSSProvider } from '../css/client-css'
import { pageToRootDocument } from '../../../client'
import { Page } from '../../../unstable-framework-support'

type Props = ComponentPropsWithoutRef<typeof Page>

export function NextRSCMakeswiftPage(props: Props) {
  const rootDocument = pageToRootDocument(props.snapshot.document)
  const rscNodes = prerenderRSCNodes(rootDocument)

  return (
    <RSCNodesProvider value={rscNodes}>
      <ClientCSSProvider>
        <CSSInjector />
        <Page {...props} />
      </ClientCSSProvider>
    </RSCNodesProvider>
  )
}
