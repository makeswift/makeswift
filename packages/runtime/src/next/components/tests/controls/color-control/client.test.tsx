/** @jest-environment jsdom */

import { Color, ColorData } from '@makeswift/controls'
import { testPageControlPropRendering } from '../page-control-prop-rendering'
import { http, HttpResponse } from 'msw'
import { server } from '../../../../../mocks/server'
import { swatchId, swatch, value, cacheData } from './fixtures'

const swatchesBaseUrl = `/api/makeswift/swatches`

describe.each([
  {
    version: 0,
    toData: (value: ColorData) => value,
  },
  {
    version: 1,
  },
])('color data v$version', ({ toData }) => {
  describe.each([
    {
      value,
      swatch,
      defaultValue: 'rgb(255,0,0)',
    },
    {
      value,
      swatch: null,
      defaultValue: 'rgb(255,0,0)',
    },
  ])(`with swatch $swatch`, ({ value, defaultValue, swatch }) => {
    test(`fetching swatch`, async () => {
      server.use(
        http.get(`${swatchesBaseUrl}/${swatchId}`, () =>
          HttpResponse.json(swatch ? { id: swatchId, ...swatch } : null),
        ),
      )

      await testPageControlPropRendering(Color({ defaultValue }), {
        toData,
        value,
        expectedRenders: swatch == null ? 1 : 2,
      })
    })

    test(`reading swatch from cache`, async () => {
      await testPageControlPropRendering(Color({ defaultValue }), {
        toData,
        value,
        cacheData: cacheData(swatch),
        expectedRenders: 1,
      })
    })
  })
})
