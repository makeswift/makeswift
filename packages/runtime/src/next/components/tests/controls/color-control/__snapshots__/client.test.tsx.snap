// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`color data v0 with swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} fetching swatch: resolvedValue 1`] = `"rgba(16, 24, 234, 0.5)"`;

exports[`color data v0 with swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} fetching swatch: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "alpha": 0.5,
                  "swatchId": "[swatch-test-id]",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "href": "test-page-href",
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`color data v0 with swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} reading swatch from cache: resolvedValue 1`] = `"rgba(16, 24, 234, 0.5)"`;

exports[`color data v0 with swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} reading swatch from cache: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "Swatch": [
        {
          "id": "[swatch-test-id]",
          "value": {
            "__typename": "Swatch",
            "hue": 238,
            "id": "[swatch-test-id]",
            "lightness": 49,
            "saturation": 87,
          },
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "alpha": 0.5,
                  "swatchId": "[swatch-test-id]",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "href": "test-page-href",
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`color data v0 with swatch null fetching swatch: resolvedValue 1`] = `"rgb(255, 0, 0)"`;

exports[`color data v0 with swatch null fetching swatch: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "alpha": 0.5,
                  "swatchId": "[swatch-test-id]",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "href": "test-page-href",
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`color data v0 with swatch null reading swatch from cache: resolvedValue 1`] = `"rgb(255, 0, 0)"`;

exports[`color data v0 with swatch null reading swatch from cache: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "Swatch": [
        {
          "id": "[swatch-test-id]",
          "value": null,
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "alpha": 0.5,
                  "swatchId": "[swatch-test-id]",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "href": "test-page-href",
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`color data v1 with swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} fetching swatch: resolvedValue 1`] = `"rgba(16, 24, 234, 0.5)"`;

exports[`color data v1 with swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} fetching swatch: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "@@makeswift/type": "color::v1",
                  "alpha": 0.5,
                  "swatchId": "[swatch-test-id]",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "href": "test-page-href",
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`color data v1 with swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} reading swatch from cache: resolvedValue 1`] = `"rgba(16, 24, 234, 0.5)"`;

exports[`color data v1 with swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} reading swatch from cache: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "Swatch": [
        {
          "id": "[swatch-test-id]",
          "value": {
            "__typename": "Swatch",
            "hue": 238,
            "id": "[swatch-test-id]",
            "lightness": 49,
            "saturation": 87,
          },
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "@@makeswift/type": "color::v1",
                  "alpha": 0.5,
                  "swatchId": "[swatch-test-id]",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "href": "test-page-href",
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`color data v1 with swatch null fetching swatch: resolvedValue 1`] = `"rgb(255, 0, 0)"`;

exports[`color data v1 with swatch null fetching swatch: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "@@makeswift/type": "color::v1",
                  "alpha": 0.5,
                  "swatchId": "[swatch-test-id]",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "href": "test-page-href",
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`color data v1 with swatch null reading swatch from cache: resolvedValue 1`] = `"rgb(255, 0, 0)"`;

exports[`color data v1 with swatch null reading swatch from cache: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "Swatch": [
        {
          "id": "[swatch-test-id]",
          "value": null,
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "@@makeswift/type": "color::v1",
                  "alpha": 0.5,
                  "swatchId": "[swatch-test-id]",
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "href": "test-page-href",
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;
